### PR TITLE
Hashing using chaining with and without replacement.

### LINKED ISSUE #2015: Hashing using chaining
Hashing using chaining with and without replacement in cpp.

The program creates a symbol table containing attribute_name and value pair as used by compiler. And includes functions like
1. Search key
2. Delete key
3. Insert key
4. Update key

Please assign this task to me.

### PR DESCRIPTION
**Pull request by utkarshanehe**
Fixes #2015

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- New file is added for hashing using chaining.
- 

#### Languages Used:
- cpp

#### Files Added:
- hashing_with_chaining.cpp

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

**Comment by raghav-dalmia** [comment]
1. Your code has some errors. Run it in your local machine.
2. Remove the second file.
3. Squash your commits.

**Comment by utkarshanehe** [comment]
Hello Sir,
         I modified the file and I committed the final changes. Could you
please review my PR #2348

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+*/
+
+#include <iostream>
+using namespace std;
+
+class symbolTableEntry
+{
+	string name,attr;
+	symbolTableEntry *next;
+public:
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: inconsistency in indentation

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+		attr="NULL";
+		next=NULL;
+	}
+	friend class symbolTable;
+};
+
+//Symbol table is basically array of objects of symbolTableEntry type
+class symbolTable
+{
+	int size,maxSize;	//Size = No. of elements in array
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: Move comments above the statement

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+//Symbol table is basically array of objects of symbolTableEntry type
+class symbolTable
+{
+	int size,maxSize;	//Size = No. of elements in array 
+						//maxSize = MAX allowed elements in table
+	symbolTableEntry table[2000];
+public:
+	symbolTable()
+	{
+		size=0;
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: Add space around the operator

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+		maxSize=0;
+	}
+
+	void traverse()		//To display all symbol table entries
+	{
+		symbolTableEntry *n;
+		cout<<"\ni  Name"<<"  Attr  "<<endl;
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: Add space around the operator

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+		for(int i=0;i<size;i++)
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: Add space around the operator

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+		{
+			n=table[i].next;
+			
+			cout<<"\n"<<i<<"  "<<table[i].name<<"  "<<table[i].attr;
+			while(n!=NULL)
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: Add space around the operator

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+			{
+				cout<<"\n"<<i<<"  "<<n->name<<"  "<<n->attr;
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: Add space around the operator

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+		symbolTableEntry *n;
+		string name;
+		int key,hash;
+		
+		cout<<"\nEnter attribute name : ";
+		cin>>name;
+		
+		//Find hash value for this key
+		key=0;
+		for(int i=0;i<name.length();i++)
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: Add space around the operator

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 9]
```diff
@@ -0,0 +1,447 @@
...
+	{
+		symbolTableEntry *n;
+		string name;
+		int key,hash;
+		
+		cout<<"\nEnter attribute name : ";
+		cin>>name;
+		
+		//Find hash value for this key
+		key=0;
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: Add space around the operator

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 10]
```diff
@@ -0,0 +1,447 @@
...
+			while(n!=NULL)
+			{
+				if(n->name.compare(name)==0)
+				{
+					cout<<"\nFOUND\n";	
+					break;
+				}
+				n=n->next;
+			}
+			if(n==NULL)
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: Add space around the operator

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 11]
```diff
@@ -0,0 +1,447 @@
+/*This is code for hashing using chaining.
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: Use 4 space indentation in the whole file
and make the required changes in the whole file

**Comment by somya-kapoor** [comment]
@utkarshanehe  Are you working on it?

**Comment by utkarshanehe** [comment]
Yes. I will finish it today.

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 12]
```diff
@@ -0,0 +1,454 @@
...
+};
+
+//Symbol table is basically array of objects of symbolTableEntry type
+class symbolTable
+{
+    //Size  =  No. of elements in array 
+    //maxSize  =  MAX allowed elements in table	
+    int size, maxSize;							
+    symbolTableEntry table[2000];
+public:
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: Indentation

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 13]
```diff
@@ -0,0 +1,454 @@
...
+        symbolTableEntry *n;
+        cout << "\ni  Name" << "  Attr  " << endl;
+        for (int i = 0; i < size; i++)
+        {
+            n = table[i].next;
+            cout << "\n" << i << "  " << table[i].name << "  " << table[i].attr;
+            while(n != NULL)
+            {
+	            cout << "\n" << i << "  " << n -> name << "  " << n -> attr;
+		        n = n -> next;
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: inconsistency in indentation

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 14]
```diff
@@ -0,0 +1,454 @@
...
+	        }
+	    }
+    } 
+
+    void search()
+    {
+        symbolTableEntry *n;
+        string name;
+	    int key, hash;
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: inconsistency in indentation

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 15]
```diff
@@ -0,0 +1,454 @@
...
+	    cout << "\nEnter attribute name : ";
+	    cin >> name;
+		
+        //Find hash value for this key
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: inconsistency in indentation

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 16]
```diff
@@ -0,0 +1,454 @@
...
+	    key = 0;
+	    for (int i = 0; i < name.length(); i++)
+	        key = key + name[i];
+
+        //Now key contains required ASCII value of entered name.
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: inconsistency in indentation

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 17]
```diff
@@ -0,0 +1,454 @@
...
+        symbolTableEntry *n;
+        string name;
+        int key, hash;
+        cout << "\nEnter attribute name : ";
+        cin >> name;
+
+        //Find hash value for this key
+        key = 0;
+        for (int i = 0; i < name.length(); i++)
+                key = key + name[i];
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: 4 space indentation

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 18]
```diff
@@ -0,0 +1,454 @@
...
+            } 
+            cout << "\nDELETED\n";
+        }	
+        else			
+        {
+            //Entry not present in table, present in chain
+            p = &(table[hash]);
+            n = table[hash].next;
+            if(n == NULL)
+            	cout << "\nNOT FOUND";
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: inconsistency in indentation

File: Hashing_With_Chaining/hashing_with_chaining.cpp [hunk 19]
```diff
@@ -0,0 +1,454 @@
...
+	            //Required name found in chain at hash address.	
+	            if (n -> name.compare(name) == 0)	
+	            {
+	                p -> next = n -> next;
+	                delete n;
+	                cout << "\nDELETED";				
+	            }
+	            else
+	            	cout << "\nNOT FOUND";				
+			}
```
Comment (somya-kapoor) [inline Hashing_With_Chaining/hashing_with_chaining.cpp]: inconsistency in indentation

**Comment by somya-kapoor** [comment]
@utkarshanehe Are you working on it?

**Comment by somya-kapoor** [comment]
Due to no activity for more than 7 days, I am closing this issue.

**Pull request closed without merging**